torch/ao/utils/__init__.py

```diff
@@ -0,0 +1,109 @@
+    r"""Generic conversion utility, that could support different modes of
+    model transformation.
+
+    Args:
+        model: Model to be transformed. Note that the model is modified inplace.
+        mapping: Dictionary that specifies the replacement logic within the
+            model.
+        config: Configuration dictionary.
+
+    Returns:
+        model: Although the transformation is inplace, the model is returned for
+            backward compatibility.
+
+    Mapping should be a dictionary, with the following keys:
+
+        - 'sparse': Dictionary with the sparse mapping
+        - 'quantized': Dictionary with the quantized mapping
+        - 'sparse_quantized': Dictionary with the sparse quantized mapping
+
+    Configuration should have the same keys as in the mapping.
+
+    Example::
+
+        >>> mapping = {
+        ...     'sparse': {
+        ...         nn.Linear: CustomSparseLinear
+        ...     },
+        ...     'quantized': {
+        ...         nn.Linear: nn.quantized.Linear
+        ...     },
+        ...     'sparse_quantized': {
+        ...         nn.Linear: CustomSparseQuantizedLinear
+        ...     }
+        ... }
+        >>> config = {
+        ...     'sparse': {
+        ...         'seq.0.linear1': {
+        ...             'zero_block_shape': (1, 4)
+        ...         }
+        ...     },
+        ...     'quantized': {
+        ...         'seq.0.linear1': {
+        ...             'qconfig': ...
+        ...         }
+        ...     },
+        ...     'sparse_quantized': {
+        ...         'seq.0.linear1': {
+        ...             'qconfig': ...,
+        ...             'zero_block_shape': (1, 4)
+        ...         }
+        ...     }
+        ... }
+        >>> torch.ao.utils.convert(model, mapping, config)
+
+    Note::
+
+        The assertion of the conversion mode is done as follows:
+
+        1. If there is a `qconfig` in a layer and there is a mask
+           parametrization, the layer will be "Sparse Quantized"
+        2. If there is a `qconfig` in a layer, but no mask parametrization, the
+           layer will be "Quantized".
+        3. If there is no `qconfig` in a layer, but there is mask
+           parametrization, the layer will be "Sparse"
+    """
+    if mapping is None:
+        mapping = dict()
+    if config is None:
+        raise AttributeError('Currently, you have to specify the convert config')
+    mapping.setdefault('quantized', torch.quantization.get_static_quant_module_class())
+    mapping.setdefault(
+        'sparse_quantized',
+        {
+            'static': {nn.Linear, torch.ao.nn.sparse.quantized.Linear},
+            'dynamic': {nn.Linear, torch.ao.nn.sparse.quantized.dynamic.Linear},
+        })
+    if 'sparse' in config:
+        raise AttributeError('FP sparse convert is not yet supported')
+
+    quant_config = config.get('quantized', dict())
+    for mode in config.keys():
```

Comment:
Lets do this instead.  We can reuse the logic in swap_module for eager mode quantization and keep the configs for 
sparsity and quantization separate
```python


def convert(model: nn.Module, mapping: dict = None, config: dict = None, convert_custom_config_dict:dict=None) -> nn.Module:
if mapping is None:
        mapping = dict()
    if config is None:
        raise AttributeError('Currently, you have to specify the convert config')
    mapping.setdefault('quantized', torch.quantization.get_static_quant_module_class())
    mapping.setdefault(
        'sparse_quantized',
        {
            'static': {nn.Linear, torch.ao.nn.sparse.quantized.Linear},
            'dynamic': {nn.Linear, torch.ao.nn.sparse.quantized.dynamic.Linear},
        })
   if 'sparse' in config:
        raise AttributeError('FP sparse convert is not yet supported')

    if convert_custom_config_dict is None:
        convert_custom_config_dict = {}
    custom_module_class_mapping = convert_custom_config_dict.get("observed_to_quantized_custom_module_class", {})

  model = _joint_convert(model, mapping, config, custom_module_class_mapping)

 def _joint_convert(model, mapping, config, custom_module_class_mapping):
       reassign = {}
       for name, mod in module.named_children():
        # both fused modules and observed custom modules are
        # swapped as one unit
        if not isinstance(mod, _FusedModule) and \
           type(mod) not in custom_module_class_mapping:
            _joint_convert(mod, mapping, config)
            # if mod has qconfig attribute and is present in sparsifier config:
            if hasattr(mod, 'qconfig') 
                 if _module_to_path(mod) in config:
                    mapping = mapping['sparse_quantized']
                 else: 
                    mapping = mapping['quantized']
               # custom module class mapping will not work with sparsity, for now we could set it to None.
                 torch.quantization.swap_module(mod, mapping, custom_module_class_mapping)
                
            
        reassign[name] = swap_module(mod, mapping, custom_module_class_mapping)

    for key, value in reassign.items():
        module._modules[key] = value
 return module